### ISSUE TITLE
Allow a broader scope for custom rules

### ISSUE THREAD

**Issue by sesponda** [StyraInc/regal#282]
Currently, the evaluation context for each custom rule is per-file (i.e. you get the AST for a single file).

We'd like to code a custom rule in the form of "_if your package is name `alfa`, then your code must have a rule named `export_rule`_"  (more context on _why_ we do this below at the end).

With a per-file scope is not possible to encode this, as two files can contribute code to the same package and trigger a false positive. Example:

File`example1.rego`:

``` 
package alfa

export_rule = x { .... } 
.. other code...

```

File `example2.rego`:

```
package alfa

.. other code...
```

In this situation, `example2.rego` fails the control. But, if both files are part of the same bundle, then the rule is not violated. For the time being, the workaround is asking Rego developers to ensure each package's code is not scattered in multiple files.

_More context about this:_
- We are building a validation system where some teams consume policy logic encoded by other teams. 
- To support standard policy aggregation and discovery patterns, have a guidelines in terms of package names and rules that packages should always have available. 
- We enforce this in runtime and unit tests. Having a linter rule will allow to converge in Regal.

**Comment by anderseknert** [comment]
Hi @sesponda! 

And thanks for filing this. It's definitely been something I'd want to do, so I guess this is a good catalyst to get started :) As for the design, I can think of two rather different approaches here.

The first one entails sending **all** parsed policy as input for evaluation, and let the policy author deal with that massive bunch of ASTs as they see fit. This would be the most flexible choice from the POV of the policy author, as you have access to all information you could possibly need from inside policy evaluation. However, this would likely perform very poorly in large repos, and the design would not easily lend itself to parallel evaluation like how we currently do in Regal.

The second option, and this is the one I've been thinking the most about, would be to keep the scope limited to a single file, but allow a new and different _aggregate report_ to be returned from linter rules. These would not work as violations as they'd merely collect some data of interest and send it back to Regal. Once the "normal" linter rules have been evaluated, the aggregate data collected from these rules would be used as input for evaluating the aggregates.

A linter rule collecting data for an aggregate could look (very roughly!) look something like this:

```rego
# METADATA
# description: export_rule required in package alfa
package example

import future.keywords

# METADATA
# custom:
#   aggregate_type: any
aggregate contains entry if {
    path := input["package"].path

    count(path) == 2
    path[1].value == "alfa"

    some rule in input.rules
    rule.head.name == "export_rule"

    entry := report.aggregate(rego.metadata.chain(), {
    	# any data that matters here
    })
}
```

A second policy / rule would then be responsible for making sense of this data, which in the case above would be to ensure that at least one item is in the aggregate report, or perhaps even *only* one item.

Thoughts?

**Comment by sesponda** [comment]
> \[complete AST\] ....  would be the most flexible choice from the POV of the policy author, as you have access to all information you could possibly need from inside policy evaluation

Agreed 👍 

> However, this would likely perform very poorly in large repos, 

I'm curios about this. We use Rego extensively, but each team develops code that fits in <50 KB. In our case, the linter process will run in the developer machine or the CI/CD build agents, both quite powerful. In fact, the code that CD/CI systems needs to deal with requires substantially more mem/CPU (large compilation, multiple e2e tests with multiple Docker images, etc.). From this perspective, even if the AST requires a couple of extra MBs, and the build takes a couple of extra seconds, I'd gladly pay the price for the increased developer experience. I reckon this is my perspective, and other groups/companies might have a completely different context. Maybe we can accomodate both worlds? (more on this below)


> and the design would not easily lend itself to parallel evaluation like how we currently do in Regal.

(what follow it's just an idea for sparring purposes - feel free to disregard/challenge/reject)

What about if we keep everything "as is" for the default behaviour, but we add an optional argument (e.g. `-with-global-ast` which, if set to true, creates that global AST so that:
- Rules using `input["package"]` will work as usual, in parallel. 
- Rules using `input["packages"]` will receive a list of package objects. That AST build will be a one-off process, and then feeding that AST to each custom rule could happen in parallel.

_(note: in our case, if all that happened sequentially it would still be OK, Regal/OPA is pretty fast for our code size)_

If the global option is not set, the second group of rules will not run (because the input is not there). The docs could clearly state the pros/cons of enabling this option. The CLI could also emit a warning if there are rules that could not be evaluated due to the missed input. In this way, only the teams needing the complete context pay the price for it.

> A linter rule collecting data for an aggregate could look (very roughly!) look something like this:

That also works. IMHO, devs (including myself) prefer to trade off CPU/mem for simplicity. Also, even though we all recognise writing good tests is paramount, the reality is that the right incentives are not present, sometimes. So, the easier we can make it for devs to write validation/verification code, the better. I'm afraid that the additional cognitive complexity of the aggregate approach could be a deterrent.

**Comment by anderseknert** [comment]
Yep, the approach is simpler to start with, but some _very_ common operations in regular linter rules, like `walk`-traversing the whole AST looking for something like variables is going to be prohibitively expensive if one replaces `input["package"]` with `input["packages"]`... and Regal simply won't be able to cope with some repos in a reasonable time-frame. Even the 50KB you mentioned would be something like 4MB of AST JSON, and millions of nodes to traverse. We already have some hacks in place to deal with the worst [performance issues](https://github.com/open-policy-agent/opa/issues/6021) of traversing huge trees with OPA, but even with those some things are just expensive.

So my concern is mainly that the "simplicity" of this model is quickly, and possibly counter-intuitively, going to be the poorer developer experience unless people understand these limitations — adding to that cognitive complexity you mentioned. Let's hope people are smart enough to only use this for what it's intended for, and not as a shortcut to be able to run "regular" rules against the whole set of ASTs, as if they do they're likely going to have a terrible experience.

I like your idea of having normal rules work as before, and only provide `input["packages"]` to custom rules that declare this dependency. I wonder what the best way of doing that would be? I guess we could add a `custom` attribute to the package-level annotation of these rules that would be parsed by Regal before deciding on how to run the rule. A configuration option would be another alternative, although that would require that the user remembers to provide configuration for these rules. I'm open to suggestions.

OTOH, I'm not sure I agree that the cognitive overhead of dealing with an aggregate would be _that_ much worse than the suggested approach. Completing my example, I could imagine even having both the "collector" and the "report" in the same rule:

```rego
# METADATA
# description: export_rule required in package alfa
package alfa.require_export_rule

import future.keywords

# this would be evaluated in parallel like other rules
aggregate contains entry if {
    path := input["package"].path

    count(path) == 2
    path[1].value == "alfa"

    some rule in input.rules
    rule.head.name == "export_rule"

    entry := report.aggregate(rego.metadata.chain(), {})
}

# this would be evaluated once all the rules ar evaluated, using the 
# output of the aggregate rule as input
report contains violation if {
    count(input.aggregate) == 0

    violation := "alfa package must contain rule named `export_rule`"
}
```

I could even imagine having a few of these added as (optional) built-in rules under the `custom` category, which would allow expressing the most common requirements — like the example you provided — using only configuration and not custom policy.

**.regal/config.yaml**
```yaml
rules:
  custom:
    require-rule:
      rule-name: export_rule
      scope:
        type: package
        pattern: '^alfa.*$'
```
Just borrowing some design from the [naming-convention](https://docs.styra.com/regal/rules/custom/naming-convention) rule, but it could probably be simplified further for this purpose.

**Comment by sesponda** [comment]
> Let's hope people are smart enough to only use this for what it's intended for, and not as a shortcut to be able to run "regular" rules against the whole set of ASTs, as if they do they're likely going to have a terrible experience.

👍 

> I like your idea of having normal rules work as before, and only provide input["packages"] to custom rules that declare this dependency.

I'm not familiar with the engine that loads the rules and traverses the Rego files to check. I've assumed is was something like this pseudocode (warning: what follows is quite probably naive! 😅 )
```
for each $file to validate:
    $ast = AST for $file
    for each $r = (custom rule): 
         evaluate $r feeding $ast
```
.. in this model, rules asking for `packages` will simply not emit a violation (as regular Rego code, when the input is missing).
An optional parameter could ask the CLI to build the global AST model, and run a 2nd phase, where the "normal" rules will not run (again, due to missed input) and the second group would run. 

> A configuration option would be another alternative, although that would require that the user remembers to provide configuration for these rules.

Yes, this is what I had in mind. This feature should be opt-in, and something devs need to fully understand before using it it, evaluating cons/pros.

> OTOH, I'm not sure I agree that the cognitive overhead of dealing with an aggregate would be that much worse than the suggested approach.

You are right, it's not a huge difference. I'm just started with Regal, and you have more exposure with customer and other use cases, so if this is the way you opt to implement, I'd gladly use it... I don't have a strong preference. Thanks for your time 🙇

**Comment by anderseknert** [comment]
It's not naive, but pretty much how it works :) I'm just saying that if instead of doing:

```
for each $files_to_validate:
```

someone were to do:

```rego
[path, value] := walk(input["packages"])

# other operations here
```

and traverse the entire list of ASTs, they're likely in for a bad time. 

I'd suggest we start with the aggregate model as it would lend itself well to a few new rules in the `custom` category others have asked for too. If it doesn't fly with people writing custom rules, I don't mind introducing another option.

Let me know if you'd like to help with this and I'd be happy to guide you :) If not, I'll try to prioritize this, but I've got quite a long list of commitments, so it could take some time before I get there.

**Comment by sesponda** [comment]
> I'd suggest we start with the aggregate model 
👍 

> Let me know if you'd like to help with this and I'd be happy to guide you :)
Of course I'd like to help. However, I have no clue about the source code of this project and my Go experience is limited. I don't know if the knowledge gap is so huge that it will take you more effort for you to babysit/guide me, than just code it yourself 🤣 

Still, I don't mind giving it a try. Worst that could happen, it's out of my depth and I'll politely decline. How do you suggest we proceed? (not in a harry, we can continue anytime)

**Comment by anderseknert** [comment]
Awesome! Some pointers to get started in the [development](https://github.com/StyraInc/regal/blob/main/docs/development.md) docs.

The main entrypoint for evaluation is unsurprisingly in [main.rego](https://github.com/StyraInc/regal/blob/main/bundle/regal/main.rego#L30), where we [query](https://github.com/StyraInc/regal/blob/main/pkg/linter/linter.go#L154) the report rule from [Go](https://github.com/StyraInc/regal/blob/main/pkg/linter/linter.go#L383).

The `report` rule in turn uses [dynamic policy composition](https://www.styra.com/blog/dynamic-policy-composition-for-opa/) to query all rules named `report` under `data.regal.rules[category][title]` for built-in rules, and `data.custom.regal.rules[category][title]` for custom rules. We could leave that logic intact for the "report style" rules, and add a new query for aggregates.

The query could have multiple bindings, so instead of:

```
violations = data.regal.main.report
```

We could query for:

```
violations := data.regal.main.report; aggregates := data.regal.main.aggregate
```

We'd then need to add corresponding `aggregate` rules to main.rego, which instead of searching for `report` rules would search for `aggregate` rules in each package.

Going back to the example, now slightly modified:

```rego
# METADATA
# description: export_rule required in package alfa
package custom.regal.rules.acmecorp["require-export-rule"]

import future.keywords

# this would be queried from regal.main.aggregate
aggregate contains entry if {
    path := input["package"].path

    count(path) == 2
    path[1].value == "alfa"

    some rule in input.rules
    rule.head.name == "export_rule"

    entry := report.aggregate(rego.metadata.chain(), {})
}

# this would get queried even in the first binding, i.e regal.main.report
# but since input.aggregate is never going to be defined, it'll never render
# a report... alternatively, we could call this aggregate_report or
# something like that to differentiate it?
report contains violation if {
    count(input.aggregate) == 0

    # second argument could contain location or other metadata
    # but that's not relevant in this example since there's no location
    # for a missing rule :)
    violation := result.fail(rego.metadata.chain(), {})
}
```

1. Create the `report.aggregate` function. This would work similarly to `report.fail` in that it would format some metadata, for inclusion in the report. But importantly, it would need to include some attributes useful in the aggregate report later. One thing that I can think of is the name of the file (i.e. `input.regal.file.name`) as we'll likely need that for accurate reporting later. But it could be any arbitrary data that is useful for the report. 
1. Extend the [Report struct](https://github.com/StyraInc/regal/blob/main/pkg/report/report.go#L38) to include an `Aggregates` field, similar to `Violations`.
1. Update [resultSetToReport](https://github.com/StyraInc/regal/blob/main/pkg/linter/linter.go#L487) to accomodate the new binding, and have it mapped to the `Aggregates` struct.
1. Once we [come back](https://github.com/StyraInc/regal/blob/main/pkg/linter/linter.go#L209) from evaluation, we'll need to inspect the report. If we have any aggregates in the report, we'll fire off a new evaluation, this time querying Rego using that as input. The violations that come back now would simply be appended to the `Violations` array, and then be sent to the reporter as we already do.

That's it, I think 😅 But I'm sure I've missed a few details that will become apparent in the implementation process. We'll figure it out, and I'd be happy to take over should you want me to. It was really useful to me just outlining the steps above!

**Comment by sesponda** [comment]
Hi @anderseknert a quick update:
- I got something working. However, I'm not convinced about the way I'm firing the 2nd evaluation (have to avoid wasting recomputation, duplicate violations, etc.) so I'm rewriting it to only collect aggregates in the first pass, and then run the eval cycle.
- Then, I'll fix/write unit tests.

To be continued some time next week.

**Comment by anderseknert** [comment]
Fantastic! No stress though. Looking forward to writing some custom rules using this 🤩

**Comment by anderseknert** [comment]
One thing that just struck me is this — what do we do with an aggregate rule if a user runs:

```
regal lint policy/my_policy.rego
```

?

I think it should be, and for later editor integrations _must_ be, OK to scan a single file for linting. I haven't looked at it, but it's probably easy to just check the number of modules lined up for linting, and if only one, modify the query to not fire off the aggregate rules, or what do you think? Perhaps you thought of that already :)

**Comment by sesponda** [comment]
I created a preliminary PR, see https://github.com/StyraInc/regal/pull/306. Should we continue there or here?
Meanwhile, I'll answer below:

> I think it should be, and for later editor integrations must be, OK to scan a single file for linting

Agreed. 

> modify the query to not fire off the aggregate rules, or what do you think? Perhaps you thought of that already :)

You'll see that what I've implemented first fires up only the `aggregates = data.regal.main.aggregate` eval for the input files, so if there are no aggregates rules, I'd expect this would be a near-NOP in terms of performance (from the users perspective, I mean perceived performance). I'd guess that when we parse all the input files (`InputFromPaths`?) we could detect if no aggregate rule is present and, in such a case, set a boolean to skip the `collectAggregates` exec?

I don't know the codebase (and Go) enough to determine if this optimisation is worth doing, or it would be premature/overkill. Feel free to suggest any changes (including this optimisation) in the PR.

_PS: while I want to contribute something that is high quality, if we spot a "wish-to-have", I'd be inclined to focus on the "must-haves" only, to optimise for completion (i.e. avoid long PR cycles, too much WIP, etc.)_ ⚖️

**Comment by anderseknert** [comment]
> PS: while I want to contribute something that is high quality, if we spot a "wish-to-have", I'd be inclined to focus on the "must-haves" only, to optimise for completion (i.e. avoid long PR cycles, too much WIP, etc.) ⚖️

Of course! Looks like a good start to me! Saving optimizations for later is a sound approach 👍 

I commented some on the PR, so let's continue there.

**Comment by anderseknert** [comment]
Resolved with #323. Will create some follow-up issues for related improvements.
